src/main/cpp/src/NativeParquetJni.cpp

```diff
@@ -0,0 +1,612 @@
+      //    always be present.
+      //  * The value field encodes the map's value type and repetition. This field can be required,
+      //    optional, or omitted.
+      //
+      // It is required that the repeated group of key-value pairs is named key_value and that its
+      // fields are named key and value. However, these names may not be used in existing data and
+      // should not be enforced as errors when reading.
+      //
+      // Some existing data incorrectly used MAP_KEY_VALUE in place of MAP. For backward-compatibility,
+      // a group annotated with MAP_KEY_VALUE that is not contained by a MAP-annotated group should be
+      // handled as a MAP-annotated group.
+
+      // Parquet says that the map's value is optional, but Spark looks like it would throw an exception
+      // if it ever actually saw that in practice, so we should too.
+      CUDF_FUNC_RANGE();
+      // The maps are sorted so we can compress the tree...
+      // These are the outputs of the computation
+      std::map<int, int, std::less<int>> chunk_map;
+      std::map<int, int, std::less<int>> schema_map;
+      std::map<int, int, std::less<int>> num_children_map;
+      // Start off with 0 children in the root, will add more as we go
+      schema_map[0] = 0;
+      num_children_map[0] = 0;
+
+      // num_children_stack and tree_stack hold the current state as we walk though schema
+      std::vector<int> num_children_stack;
+      std::vector<column_pruner*> tree_stack;
+      tree_stack.push_back(this);
+      num_children_stack.push_back(schema[0].num_children);
+
+      uint64_t chunk_index = 0;
+      // We are skipping over the first entry in the schema because it is always the root entry, and
+      //  we already processed it
+      for (uint64_t schema_index = 1; schema_index < schema.size(); schema_index++) {
+        auto schema_item = schema[schema_index];
+        // num_children is optional, but is supposed to be set for non-leaf nodes. That said leaf nodes
+        // will have 0 children so we can just default to that.
+        int num_children = 0;
+        if (schema_item.__isset.num_children) {
+          num_children = schema_item.num_children;
+        }
+        std::string name;
+        if (ignore_case) {
+          name = unicode_to_lower(schema_item.name);
+        } else {
+          name = schema_item.name;
+        }
+        column_pruner * found = nullptr;
+        if (tree_stack.back() != nullptr) {
+          // tree_stack can have a nullptr in it if the scheam we are looking through
+          // has an entry that does not match the tree
+          auto found_it = tree_stack.back()->children.find(name);
+          if (found_it != tree_stack.back()->children.end()) {
+            found = &(found_it->second);
+            int parent_mapped_schema_index = tree_stack.back()->s_id;
+            num_children_map[parent_mapped_schema_index]++;
+          }
+        }
+
+        if (schema_item.__isset.type) {
```

Comment:
it seems like code inside of this if statement (other than `chunk_index++`) executes if `found != nullptr`. Could it be moved over to line 222 where we set `found` to be not null?